# Final Project
### Dataset:
https://archive.ics.uci.edu/ml/datasets/adult  
Classifying: >50K, <=50K.  

**age:** continuous.  
**workclass:** Private, Self-emp-not-inc, Self-emp-inc, Federal-gov, Local-gov, State-gov, Without-pay, Never-worked.  
**fnlwgt:** continuous.  
**education:** Bachelors, Some-college, 11th, HS-grad, Prof-school, Assoc-acdm, Assoc-voc, 9th, 7th-8th, 12th, Masters, 1st-4th, 10th, Doctorate, 5th-6th, Preschool.  
**education-num:** continuous.  
**marital-status:** Married-civ-spouse, Divorced, Never-married, Separated, Widowed, Married-spouse-absent, Married-AF-spouse.  
**occupation:** Tech-support, Craft-repair, Other-service, Sales, Exec-managerial, Prof-specialty, Handlers-cleaners, Machine-op-inspct, Adm-clerical, Farming-fishing, Transport-moving, Priv-house-serv, Protective-serv, Armed-Forces.  
**relationship:** Wife, Own-child, Husband, Not-in-family, Other-relative, Unmarried.  
**race:** White, Asian-Pac-Islander, Amer-Indian-Eskimo, Other, Black.  
**sex:** Female, Male.  
**capital-gain:** continuous.  
**capital-loss:** continuous.  
**hours-per-week:** continuous.  
**native-country:** United-States, Cambodia, England, Puerto-Rico, Canada, Germany, Outlying-US(Guam-USVI-etc), India, Japan, Greece, South, China, Cuba, Iran, Honduras, Philippines, Italy, Poland, Jamaica, Vietnam, Mexico, Portugal, Ireland, France, Dominican-Republic, Laos, Ecuador, Taiwan, Haiti, Columbia, Hungary, Guatemala, Nicaragua, Scotland, Thailand, Yugoslavia, El-Salvador, Trinadad&Tobago, Peru, Hong, Holand-Netherlands.

In [24]:
# Imports
import numpy as np
import pandas as pd
import seaborn as sns
from sklearn import *
from scipy import stats

In [2]:
data = pd.read_csv("adult.csv")
data.head()

age          workclass   fnlwgt   education   education-num  \
0   39          State-gov    77516   Bachelors              13   
1   50   Self-emp-not-inc    83311   Bachelors              13   
2   38            Private   215646     HS-grad               9   
3   53            Private   234721        11th               7   
4   28            Private   338409   Bachelors              13   

        marital-status          occupation    relationship    race      sex  \
0        Never-married        Adm-clerical   Not-in-family   White     Male   
1   Married-civ-spouse     Exec-managerial         Husband   White     Male   
2             Divorced   Handlers-cleaners   Not-in-family   White     Male   
3   Married-civ-spouse   Handlers-cleaners         Husband   Black     Male   
4   Married-civ-spouse      Prof-specialty            Wife   Black   Female   

    capital-gain   capital-loss   hours-per-week  native-country  income  
0           2174              0               40   United-States   <=50K  
1              0              0               13   United-States   <=50K  
2              0              0               40   United-States   <=50K  
3              0              0               40   United-States   <=50K  
4              0              0               40            Cuba   <=50K

### Data Cleaning

In [3]:
old_data_types = data.dtypes
data.columns = data.columns.str.strip()
data = data.apply(lambda x : x.astype(str).str.strip(' .') if not pd.api.types.is_int64_dtype(x) else x)

In [4]:
value_instances = data.apply(lambda x : dict(x.value_counts()))

In [5]:
print(value_instances['marital-status'])
print(value_instances['relationship'])

{'Married-civ-spouse': 22379, 'Never-married': 16117, 'Divorced': 6633, 'Separated': 1530, 'Widowed': 1518, 'Married-spouse-absent': 628, 'Married-AF-spouse': 37}
{'Husband': 19716, 'Not-in-family': 12583, 'Own-child': 7581, 'Unmarried': 5125, 'Wife': 2331, 'Other-relative': 1506}


#### Pruning Redundant Data via Correlation Analysis
Features of interest:
* `education` and `education-num`
* `marital-status` and `relationship`

In [6]:
xtrain_edu, xtest_edu, ytrain_edu, ytest_edu = model_selection.train_test_split(np.array(data['education-num']), 
                                                                                 np.array(data['education']),
                                                                                 test_size=0.2,
                                                                                 random_state=0)

In [7]:
# https://medium.com/@outside2SDs/an-overview-of-correlation-measures-between-categorical-and-continuous-variables-4c7f85610365
clf_edu = linear_model.LogisticRegression(solver='newton-cg').fit(xtrain_edu.reshape(-1, 1), ytrain_edu)

In [8]:
predict_edu = clf_edu.predict(xtest_edu.reshape(-1, 1))
print(predict_edu)
print(f"Accuracy: {metrics.accuracy_score(ytest_edu, predict_edu)}")

['HS-grad' 'Doctorate' 'Bachelors' ... 'Some-college' 'Assoc-voc'
 'Bachelors']
Accuracy: 1.0


In [69]:
display(pd.crosstab(data['marital-status'], data['relationship']))
chi2_mr, p_mr, df_mr, expected_mr = stats.chi2_contingency(pd.crosstab(data['marital-status'], data['relationship']))
print(f"Chi2 Value: {chi2_mr}")

relationship           Husband  Not-in-family  Other-relative  Own-child  \
marital-status                                                             
Divorced                     0           3628             181        455   
Married-AF-spouse           12              0               1          1   
Married-civ-spouse       19704             23             201        143   
Married-spouse-absent        0            330              54         61   
Never-married                0           7114             920       6750   
Separated                    0            637              79        146   
Widowed                      0            851              70         25   

relationship           Unmarried  Wife  
marital-status                          
Divorced                    2369     0  
Married-AF-spouse              0    23  
Married-civ-spouse             0  2308  
Married-spouse-absent        183     0  
Never-married               1333     0  
Separated                    668     0  
Widowed                      572     0

Chi2 Value: 58195.24158415406


**Chi Square for Pruning**  
* Null Hypothesis: `marital-status` and `relationship` are not correlated
* Significance Level: `0.05`
* DoF: (7-1)(6-1) = `30`
* Critical Value: `43.77`  

*Result:*  
* Chi2 Value: `58195.24158415406`
* In order to fail to reject the null hypothesis, we need our Chi2 Value to be less than `43.77`. However, after performing Chi2 analysis, we get `58195.24158415406`, which is much larger than the `43.77`. This means that we reject our null hypothesis that `marital-status` and `relationship` are most definitely correlated, and we can prune one of the features.

In [18]:
# dropping redundant features
data.drop(columns='education')

age         workclass  fnlwgt  education      marital-status  \
0       39         State-gov   77516  Bachelors       Never-married   
1       50  Self-emp-not-inc   83311  Bachelors  Married-civ-spouse   
2       38           Private  215646    HS-grad            Divorced   
3       53           Private  234721       11th  Married-civ-spouse   
4       28           Private  338409  Bachelors  Married-civ-spouse   
...    ...               ...     ...        ...                 ...   
48837   39           Private  215419  Bachelors            Divorced   
48838   64                 ?  321403    HS-grad             Widowed   
48839   38           Private  374983  Bachelors  Married-civ-spouse   
48840   44           Private   83891  Bachelors            Divorced   
48841   35      Self-emp-inc  182148  Bachelors  Married-civ-spouse   

              occupation    relationship                race     sex  \
0           Adm-clerical   Not-in-family               White    Male   
1        Exec-managerial         Husband               White    Male   
2      Handlers-cleaners   Not-in-family               White    Male   
3      Handlers-cleaners         Husband               Black    Male   
4         Prof-specialty            Wife               Black  Female   
...                  ...             ...                 ...     ...   
48837     Prof-specialty   Not-in-family               White  Female   
48838                  ?  Other-relative               Black    Male   
48839     Prof-specialty         Husband               White    Male   
48840       Adm-clerical       Own-child  Asian-Pac-Islander    Male   
48841    Exec-managerial         Husband               White    Male   

       capital-gain  capital-loss  hours-per-week native-country income  
0              2174             0              40  United-States  <=50K  
1                 0             0              13  United-States  <=50K  
2                 0             0              40  United-States  <=50K  
3                 0             0              40  United-States  <=50K  
4                 0             0              40           Cuba  <=50K  
...             ...           ...             ...            ...    ...  
48837             0             0              36  United-States  <=50K  
48838             0             0              40  United-States  <=50K  
48839             0             0              50  United-States  <=50K  
48840          5455             0              40  United-States  <=50K  
48841             0             0              60  United-States   >50K  

[48842 rows x 14 columns]